File: Sources/Features/GitChanges/Views/DiffView.swift
// MARK: - DiffView
// Side-by-side (split) diff view — old file on the left, new on the right.
// macOS 14+, Swift 5.10

import SwiftUI

/// Side-by-side split diff view.
///
/// Renders git diff hunks in two equal-width columns:
/// - **Left** — old file: context lines + deleted lines (red background).
/// - **Right** — new file: context lines + added lines (green background).
///
/// Deletion/addition pairs in the same change block are aligned row-by-row
/// for easy visual comparison. Hunk headers span the full width.
struct DiffView: View {

    let hunks: [GitDiffHunk]

    var body: some View {
        let rows = buildRows(from: hunks)
        return ScrollView(.vertical) {
            LazyVStack(alignment: .leading, spacing: 0) {
                ForEach(rows) { row in
                    rowView(row)
                }
            }
        }
        .background(DSColor.surfaceBase)
    }

    // MARK: - Row Views

    @ViewBuilder
    private func rowView(_ row: SplitRow) -> some View {
        switch row.kind {
        case .hunkHeader(let text):
            Text(text)
                .font(DSFont.terminal(size: 11))
                .foregroundStyle(DSColor.textMuted)
                .padding(.horizontal, DSSpacing.sm)
                .frame(maxWidth: .infinity, alignment: .leading)
                .frame(height: DSLayout.diffLineHeight)
                .background(DSColor.surfaceOverlay)

        case .content(let left, let right):
            HStack(spacing: 0) {
                sideCell(left)
                    .frame(maxWidth: .infinity)
                    .clipped()
                Rectangle()
                    .fill(DSColor.diffGutter.opacity(0.25))
                    .frame(width: 1)
                sideCell(right)
                    .frame(maxWidth: .infinity)
                    .clipped()
            }
            .frame(height: DSLayout.diffLineHeight)
        }
    }

    // MARK: - Side Cell

    private func sideCell(_ cell: SplitDiffCell?) -> some View {
        HStack(spacing: 0) {
            // Line number gutter
            Group {
                if let num = cell?.lineNumber {
                    Text("\(num)")
                } else {
                    Text("")
                }
            }
            .font(DSFont.terminal(size: 11))
            .foregroundStyle(DSColor.diffGutter)
            .frame(width: 32, alignment: .trailing)
            .padding(.trailing, 3)

            // +/- symbol
            Text(prefix(for: cell?.type))
                .font(DSFont.terminal(size: 12))
                .foregroundStyle(prefixColor(for: cell?.type))
                .frame(width: 12)

            // Line content
            Text(cell?.content ?? "")
                .font(DSFont.terminal(size: 12))
                .foregroundStyle(DSColor.textPrimary)
                .lineLimit(1)
                .textSelection(.enabled)
                .frame(maxWidth: .infinity, alignment: .leading)

            Spacer(minLength: 4)
        }
        .frame(maxWidth: .infinity)
        .background(cellBackground(for: cell))
    }

    // MARK: - Helpers

    private func prefix(for type: DiffLineType?) -> String {
        switch type {
        case .addition: "+"
        case .deletion: "-"
        case .context, .none: " "
        }
    }

    private func prefixColor(for type: DiffLineType?) -> Color {
        switch type {
        case .addition: DSColor.gitAdded
        case .deletion: DSColor.gitDeleted
        case .context, .none: DSColor.textMuted
        }
    }

    private func cellBackground(for cell: SplitDiffCell?) -> Color {
        guard let cell else {
            // Empty placeholder — subtle tint so the gutter/content area
            // is visually distinct from context lines.
            return DSColor.diffGutter.opacity(0.06)
        }
        switch cell.type {
        case .addition: return DSColor.diffAddedBg
        case .deletion: return DSColor.diffDeletedBg
        case .context:  return .clear
        }
    }

    // MARK: - Row Builder

    /// Converts raw diff hunks into flat `SplitRow` array for rendering.
    ///
    /// Change blocks (consecutive deletions + additions) are aligned:
    /// each deletion is paired with the corresponding addition on the same row.
    /// Extra deletions get an empty right cell; extra additions get an empty left cell.
    private func buildRows(from hunks: [GitDiffHunk]) -> [SplitRow] {
        var rows: [SplitRow] = []
        var nextId = 0

        for hunk in hunks {
            rows.append(SplitRow(id: nextId, kind: .hunkHeader(hunk.header)))
            nextId += 1

            var i = 0
            let lines = hunk.lines

            while i < lines.count {
                switch lines[i].type {

                case .context:
                    let line = lines[i]
                    rows.append(SplitRow(id: nextId, kind: .content(
                        left:  SplitDiffCell(lineNumber: line.oldLineNumber, content: line.content, type: .context),
                        right: SplitDiffCell(lineNumber: line.newLineNumber, content: line.content, type: .context)
                    )))
                    nextId += 1
                    i += 1

                case .deletion:
                    // Collect consecutive deletions…
                    var deletions: [GitDiffLine] = []
                    while i < lines.count && lines[i].type == .deletion {
                        deletions.append(lines[i])
                        i += 1
                    }
                    // …then any additions that immediately follow (change block).
                    var additions: [GitDiffLine] = []
                    while i < lines.count && lines[i].type == .addition {
                        additions.append(lines[i])
                        i += 1
                    }
                    // Pair them row-by-row; extras get a nil on the opposite side.
                    for j in 0..<max(deletions.count, additions.count) {
                        let del = j < deletions.count ? deletions[j] : nil
                        let add = j < additions.count ? additions[j] : nil
                        rows.append(SplitRow(id: nextId, kind: .content(
                            left:  del.map { SplitDiffCell(lineNumber: $0.oldLineNumber, content: $0.content, type: .deletion) },
                            right: add.map { SplitDiffCell(lineNumber: $0.newLineNumber, content: $0.content, type: .addition) }
                        )))
                        nextId += 1
                    }

                case .addition:
                    // Standalone addition (not immediately preceded by a deletion block).
                    let line = lines[i]
                    rows.append(SplitRow(id: nextId, kind: .content(
                        left:  nil,
                        right: SplitDiffCell(lineNumber: line.newLineNumber, content: line.content, type: .addition)
                    )))
                    nextId += 1
                    i += 1
                }
            }
        }

        return rows
    }
}

// MARK: - Private Models

private struct SplitRow: Identifiable {
    let id: Int
    enum Kind {
        case hunkHeader(String)
        case content(left: SplitDiffCell?, right: SplitDiffCell?)
    }
    let kind: Kind
}

private struct SplitDiffCell {
    let lineNumber: Int?
    let content: String
    let type: DiffLineType
}

File: desktopApp/src/main/kotlin/studio/vibe/desktop/ui/DiffView.kt
@file:OptIn(kotlin.uuid.ExperimentalUuidApi::class)
@file:Suppress("DEPRECATION")

package studio.vibe.desktop.ui

import androidx.compose.foundation.ScrollbarAdapter
import androidx.compose.foundation.VerticalScrollbar
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import studio.vibe.desktop.ui.theme.DSColor
import studio.vibe.desktop.ui.theme.DSFont
import studio.vibe.desktop.ui.theme.DSSpacing

// ── Constants ─────────────────────────────────────────────────────────────────

private val LINE_NUMBER_WIDTH = 48.dp
private val PREFIX_WIDTH = 16.dp
private val DIFF_LINE_HEIGHT = 20.dp
private val GUTTER_DIVIDER_WIDTH = 1.dp

// ── Public entry point ────────────────────────────────────────────────────────

/**
 * Side-by-side split diff viewer.
 *
 * Accepts a raw unified diff string (output of `git diff`) and renders it in
 * two equal-width columns — old file on the left, new file on the right.
 * Deletion/addition pairs within the same change block are row-aligned so the
 * reader can compare them at a glance.  Hunk headers span the full width.
 *
 * @param diffText Raw unified diff text (may be empty or a "binary file" stub).
 * @param modifier Optional layout modifier applied to the outer container.
 */
@Composable
fun DiffView(
    diffText: String,
    modifier: Modifier = Modifier,
) {
    val rows = remember(diffText) { parseDiff(diffText) }
    val listState = rememberLazyListState()

    Box(modifier = modifier.background(DSColor.surfaceBase)) {
        if (rows.isEmpty()) {
            EmptyDiffPlaceholder()
        } else {
            LazyColumn(
                state = listState,
                modifier = Modifier.fillMaxSize(),
            ) {
                itemsIndexed(rows, key = { index, _ -> index }) { _, row ->
                    DiffRowView(row)
                }
            }
            VerticalScrollbar(
                adapter = ScrollbarAdapter(listState),
                modifier = Modifier
                    .align(Alignment.CenterEnd)
                    .fillMaxHeight()
                    .padding(end = 2.dp),
            )
        }
    }
}

// ── Row rendering ─────────────────────────────────────────────────────────────

@Composable
private fun DiffRowView(row: SplitRow) {
    when (val kind = row.kind) {
        is SplitRow.Kind.HunkHeader -> HunkHeaderRow(kind.text)
        is SplitRow.Kind.Content -> ContentRow(kind.left, kind.right)
    }
}

@Composable
private fun HunkHeaderRow(text: String) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(DIFF_LINE_HEIGHT)
            .background(DSColor.surfaceOverlay)
            .padding(horizontal = DSSpacing.sm),
        contentAlignment = Alignment.CenterStart,
    ) {
        Text(
            text = text,
            style = DSFont.monoSmall,
            color = DSColor.textMuted,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
private fun ContentRow(left: SplitDiffCell?, right: SplitDiffCell?) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(DIFF_LINE_HEIGHT),
    ) {
        SideCell(
            cell = left,
            modifier = Modifier.weight(1f),
        )
        // Centre divider
        Box(
            Modifier
                .width(GUTTER_DIVIDER_WIDTH)
                .fillMaxHeight()
                .background(DSColor.diffGutter.copy(alpha = 0.25f)),
        )
        SideCell(
            cell = right,
            modifier = Modifier.weight(1f),
        )
    }
}

@Composable
private fun SideCell(cell: SplitDiffCell?, modifier: Modifier = Modifier) {
    val bgColor = when (cell?.type) {
        DiffCellType.ADDITION -> DSColor.diffAddedBg
        DiffCellType.DELETION -> DSColor.diffDeletedBg
        DiffCellType.CONTEXT, null -> Color.Transparent
    }
    val emptyBg = if (cell == null) DSColor.diffGutter.copy(alpha = 0.06f) else Color.Transparent

    Row(
        modifier = modifier
            .fillMaxHeight()
            .background(if (cell != null) bgColor else emptyBg),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Line number
        Text(
            text = cell?.lineNumber?.toString() ?: "",
            style = DSFont.monoSmall.copy(fontSize = 11.sp),
            color = DSColor.diffGutter,
            modifier = Modifier
                .width(LINE_NUMBER_WIDTH)
                .padding(end = DSSpacing.xxs),
            maxLines = 1,
        )

        // +/- prefix symbol
        val prefixText = when (cell?.type) {
            DiffCellType.ADDITION -> "+"
            DiffCellType.DELETION -> "-"
            DiffCellType.CONTEXT, null -> " "
        }
        val prefixColor = when (cell?.type) {
            DiffCellType.ADDITION -> DSColor.gitAdded
            DiffCellType.DELETION -> DSColor.gitDeleted
            DiffCellType.CONTEXT, null -> DSColor.textMuted
        }
        Text(
            text = prefixText,
            style = DSFont.monoSmall.copy(fontSize = 12.sp),
            color = prefixColor,
            modifier = Modifier.width(PREFIX_WIDTH),
            maxLines = 1,
        )

        // Line content
        val contentColor = when (cell?.type) {
            DiffCellType.ADDITION -> DSColor.gitAdded
            DiffCellType.DELETION -> DSColor.gitDeleted
            DiffCellType.CONTEXT -> DSColor.textSecondary
            null -> Color.Transparent
        }
        Text(
            text = cell?.content ?: "",
            style = DSFont.monoSmall.copy(fontSize = 12.sp),
            color = contentColor,
            maxLines = 1,
            overflow = TextOverflow.Clip,
            modifier = Modifier
                .weight(1f)
                .padding(end = DSSpacing.xs),
        )
    }
}

@Composable
private fun EmptyDiffPlaceholder() {
    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        Text(
            text = "No changes",
            style = DSFont.sidebarItem,
            color = DSColor.textMuted,
        )
    }
}

// ── Unified diff parser ───────────────────────────────────────────────────────

/**
 * Parses a raw unified diff string into a flat list of [SplitRow] values
 * suitable for direct rendering in a [LazyColumn].
 *
 * Algorithm:
 * 1. Skip file header lines (`---` / `+++`).
 * 2. Hunk header lines (`@@…`) become full-width [SplitRow.Kind.HunkHeader] rows.
 * 3. Inside a hunk, consecutive deletions followed by consecutive additions are
 *    paired row-by-row so both sides align visually.  Unpaired deletions get a
 *    null right cell; unpaired additions get a null left cell.
 * 4. Context lines (space-prefixed) appear on both sides with matching line numbers.
 */
private fun parseDiff(diffText: String): List<SplitRow> {
    if (diffText.isBlank()) return emptyList()

    val rawLines = diffText.lines()
    // Convert raw text into typed DiffInputLine values
    val inputLines = mutableListOf<DiffInputLine>()
    var oldLine = 0
    var newLine = 0

    for (raw in rawLines) {
        when {
            raw.startsWith("--- ") || raw.startsWith("+++ ") -> {
                // File header — skip entirely (not rendered)
                continue
            }

            raw.startsWith("@@ ") -> {
                // Parse hunk header to extract start line numbers
                // Format: @@ -<old_start>[,<count>] +<new_start>[,<count>] @@
                val match = Regex("""^@@\s+-(\d+)(?:,\d+)?\s+\+(\d+)(?:,\d+)?\s+@@""").find(raw)
                if (match != null) {
                    oldLine = match.groupValues[1].toIntOrNull() ?: 1
                    newLine = match.groupValues[2].toIntOrNull() ?: 1
                }
                inputLines += DiffInputLine.HunkHeader(raw)
            }

            raw.startsWith("-") -> {
                inputLines += DiffInputLine.Deletion(raw.drop(1), oldLine)
                oldLine++
            }

            raw.startsWith("+") -> {
                inputLines += DiffInputLine.Addition(raw.drop(1), newLine)
                newLine++
            }

            raw.startsWith(" ") || raw.isEmpty() -> {
                // Context line (space prefix or empty — some diffs omit the space on blank lines)
                inputLines += DiffInputLine.Context(
                    content = if (raw.startsWith(" ")) raw.drop(1) else raw,
                    oldLineNumber = oldLine,
                    newLineNumber = newLine,
                )
                oldLine++
                newLine++
            }

            else -> {
                // Unknown — ignore (diff meta lines like "diff --git", "index …", "new file mode")
                continue
            }
        }
    }

    return buildSplitRows(inputLines)
}

private fun buildSplitRows(inputLines: List<DiffInputLine>): List<SplitRow> {
    val rows = mutableListOf<SplitRow>()
    var i = 0

    while (i < inputLines.size) {
        when (val line = inputLines[i]) {
            is DiffInputLine.HunkHeader -> {
                rows += SplitRow(SplitRow.Kind.HunkHeader(line.text))
                i++
            }

            is DiffInputLine.Context -> {
                rows += SplitRow(
                    SplitRow.Kind.Content(
                        left = SplitDiffCell(line.oldLineNumber, line.content, DiffCellType.CONTEXT),
                        right = SplitDiffCell(line.newLineNumber, line.content, DiffCellType.CONTEXT),
                    ),
                )
                i++
            }

            is DiffInputLine.Deletion -> {
                // Collect all consecutive deletions…
                val deletions = mutableListOf<DiffInputLine.Deletion>()
                while (i < inputLines.size && inputLines[i] is DiffInputLine.Deletion) {
                    deletions += inputLines[i] as DiffInputLine.Deletion
                    i++
                }
                // …then any additions that immediately follow (same change block)
                val additions = mutableListOf<DiffInputLine.Addition>()
                while (i < inputLines.size && inputLines[i] is DiffInputLine.Addition) {
                    additions += inputLines[i] as DiffInputLine.Addition
                    i++
                }
                // Pair them row-by-row; extras get a null on the opposite side
                val pairCount = maxOf(deletions.size, additions.size)
                for (j in 0 until pairCount) {
                    val del = deletions.getOrNull(j)
                    val add = additions.getOrNull(j)
                    rows += SplitRow(
                        SplitRow.Kind.Content(
                            left = del?.let { SplitDiffCell(it.lineNumber, it.content, DiffCellType.DELETION) },
                            right = add?.let { SplitDiffCell(it.lineNumber, it.content, DiffCellType.ADDITION) },
                        ),
                    )
                }
            }

            is DiffInputLine.Addition -> {
                // Standalone addition (not preceded by deletions in the same block)
                rows += SplitRow(
                    SplitRow.Kind.Content(
                        left = null,
                        right = SplitDiffCell(line.lineNumber, line.content, DiffCellType.ADDITION),
                    ),
                )
                i++
            }
        }
    }

    return rows
}

// ── Private models ────────────────────────────────────────────────────────────

/** Intermediate representation of a single line from the raw diff text. */
private sealed interface DiffInputLine {
    data class HunkHeader(val text: String) : DiffInputLine
    data class Context(val content: String, val oldLineNumber: Int, val newLineNumber: Int) : DiffInputLine
    data class Deletion(val content: String, val lineNumber: Int) : DiffInputLine
    data class Addition(val content: String, val lineNumber: Int) : DiffInputLine
}

/** A single cell on one side of the split view. */
private data class SplitDiffCell(
    val lineNumber: Int?,
    val content: String,
    val type: DiffCellType,
)

private enum class DiffCellType { CONTEXT, ADDITION, DELETION }

/** One rendered row in the diff — either a full-width hunk header or a paired content row. */
private data class SplitRow(val kind: Kind) {
    sealed interface Kind {
        data class HunkHeader(val text: String) : Kind
        data class Content(val left: SplitDiffCell?, val right: SplitDiffCell?) : Kind
    }
}
